### PR TITLE
hip : skip fully masked KV tiles in WMMA Flash Attention

### PR DESCRIPTION
**Pull request by moristeven477-ship-it**
## What changed

When server slots share a unified KV cache, the AMD WMMA Flash Attention kernel can still compute interior KV tiles that are fully masked for the current request. Following the review suggestion, this revision moves mask scanning and reduction out of the attention loop and into an auxiliary preprocessing kernel.

The new flash_attn_mask_to_KV_blocks classifies each block as fully masked, all-zero, or general. The attention kernel skips runs of fully masked blocks, omits mask loading and mask addition for all-zero runs, and keeps normal mask evaluation for general blocks. Sixteen two-bit classes pack into a single 32-bit word and are reused across attention heads.

This path is limited to synchronous, dense AMD WMMA. It covers the measured -ub 512 --kv-unified configuration, in which the existing flash_attn_mask_to_KV_max helper is not launched. The patch also includes the broadcast-mask and query-tail fixes plus additional operator tests.

On RX 7900 XTX / Ubuntu 24.04 / ROCm 7.2.1, third-request time drops by approximately 33.3% versus upstream, and by a further 10% versus V1. The gain is primarily in prefill; this is not a general decode or concurrent-server throughput claim.

The model is Qwen3.8-27B-UD-Q4_K_M, fully offloaded, with Flash Attention enabled. The workload uses 98,304 context capacity, 40,960 input / 64 output tokens, and two unified-cache slots requested sequentially in 0/1/0 order. Ubatch is 512, with no prompt cache reuse and no speculative decoding.

| KV cache | Upstream | V2 | 95% CI for relative time change, V2 vs upstream |
| --- | ---: | ---: | ---: |
| F16 | 77.8 s | 51.9 s | [-33.447%, -33.225%] |
| Q8_0 | 80.3 s | 53.5 s | [-33.330%, -33.257%] |

Times are geometric means over six restarts per variant, covering all six U/V1/V2 execution orders on the common base c6824a9. Confidence intervals use paired log-ratios. The current source is ported to ec91ab5; the three-arm performance comparison has not been repeated on that base.

## Validation

- **Operator tests on ec91:** 4,007 passed, 0 failed, 1,323 unsupported. The extended suite adds boundary cases. Upstream and V1 each fail six of the new broadcast-mask cases that V2 fixes; these are not failures in pre-existing upstream CI tests.
- **Model checks:** fixed F16/Q8_0 perplexity matches the per-chunk reference values at printed precision, without claiming bitwise-identical internals. Sequential requests match the reference token IDs.
- **Kernel trace:** all 3,840 WMMA calls in the instrumented run had a preceding classifier launch on the same stream. Peak dispatch metadata was 5,184 bytes; this is not total extra VRAM.
- **Local HIP CI:** passed in aggregate, after one HTTP-test failure was re-run and passed and the model-download interruptions were resolved. This was not an uninterrupted all-green run. The full operator suite also relied on a separate TOPK comparator fix, which is excluded from this attention patch.

The [V2 evidence release](https://github.com/moristeven477-ship-it/llama.cpp/releases/tag/rdna3-masked-kv-v2-20260921) contains the patch, commands, hashes, raw data, and logs. The V1 artifacts do not substitute for this evidence.

- [V2 patch on ec91](https://github.com/moristeven477-ship-it/llama.cpp/releases/download/rdna3-masked-kv-v2-20260921/attention-v2-ec91.patch); [SHA256 checksums](https://github.com/moristeven477-ship-it/llama.cpp/releases/download/rdna3-masked-kv-v2-20260921/SHA256SUMS).
- Reproduction commands: [c682 performance](https://github.com/moristeven477-ship-it/llama.cpp/releases/download/rdna3-masked-kv-v2-20260921/REPRODUCE-c682.md) / [ec91 validation](https://github.com/moristeven477-ship-it/llama.cpp/releases/download/rdna3-masked-kv-v2-20260921/REPRODUCE-ec91.md).
- [Raw performance data](https://github.com/moristeven477-ship-it/llama.cpp/releases/download/rdna3-masked-kv-v2-20260921/p23-primary-native-20260920.tar.gz) and [primary-workload findings](https://github.com/moristeven477-ship-it/llama.cpp/releases/download/rdna3-masked-kv-v2-20260921/primary-FINDINGS.zh.md).
- [CI and model logs](https://github.com/moristeven477-ship-it/llama.cpp/releases/download/rdna3-masked-kv-v2-20260921/p23-ec91-ci-and-models-20260921.tar.gz), plus [CI scope](https://github.com/moristeven477-ship-it/llama.cpp/releases/download/rdna3-masked-kv-v2-20260921/ci-FINDINGS.zh.md).
- [Matched operator tests, calibration, and raw kernel traces](https://github.com/moristeven477-ship-it/llama.cpp/releases/download/rdna3-masked-kv-v2-20260921/p23-ec91-final-evidence-20260921.tar.gz).

## Credit & inspiration

Thanks to JohannesGaessler for the auxiliary-preprocessing suggestion, to DanoPTT for the V1 testing and the helper-launch diagnosis, and to alankila together with the earlier reports in #28495 and #14924. The masked-block preprocessing idea follows other backends, including Vulkan (#17186); the ec91 port retains ynankani's swizzle work (#28536).

### CODE REVIEW

**Comment by JohannesGaessler** [comment]
There already is an auxiliary kernel `flash_attn_mask_to_KV_max` that makes it possible to skip fully masked KV slices for multiple sequences. A solution for unified KV cache should look something like this. Avoid dropping a bunch of extra instructions into an inner loop as was done in this PR.

**Comment by DanoPTT** [comment]
I offered to test this on gfx1201 in #28495, so here are the numbers.

Short version: the kernel hunk gives me **+8 to +19 % prefill** depending on
depth, the extra work in the inner loop costs **nothing measurable**, and decode
is unchanged. I also found that on this configuration
`flash_attn_mask_to_KV_max` is never launched at all, which is relevant to the
review comment.

## Setup

- Radeon AI PRO R9700 (gfx1201, RDNA4), Windows 11 build 26200, ROCm runtime 10.0.0
- Model: Qwen3.8-27B-UD-Q5_K_XL, arch `qwen35`, head 256, gqa 6 — so the AMD WMMA
  path, not MFMA
- `-c 161280 -np 3 --kv-unified --no-cache-idle-slots -ub 512 -b 1024`, F16 KV,
  `--spec-type draft-mtp --spec-draft-n-max 3`, `-ngl 999`, FA on,
  `GGML_CUDA_DISABLE_GRAPHS=1`, GPU otherwise idle
- `--cache-ram 0` and `cache_prompt=false` so the measurement is the kernel and
  not the prompt cache
- Both arms are CI builds from the same workflow, same ROCm, same runtime tree;
  they differ only in the 19-line kernel hunk. I am on a fork, so the absolute
  numbers are mine, but the delta is this patch.

One deviation from your patch: the `if constexpr (!use_sparse)` guard is dropped,
because sparse-FA (#27970) is not in my tree. The test fixtures are not ported —
my `test_flash_attn_ext` signature predates `n_kv_max`/`kv_view`.

## 1. Prefill

Requests are issued **sequentially** into different slots (0,1,2,0,1,2), so at
step 0 the other slots are empty and by step 3 all three hold ~5.4k or ~11k
tokens each. Median over 4 visits per arm, arms interleaved A,B,B,A,A,B,B,A.
"sep" is how many of the 16 (A-visit, B-visit) pairs had B faster.

Prompt ~5,400 tokens:

| step | stock pp t/s | patch pp t/s | delta | sep |
| ---: | ---: | ---: | --- | --- |
| 0 | 963.2 | 966.6 | +0.4 % | 7/16 |
| 1 | 911.9 | 947.4 | +3.9 % | 15/16 |
| 2 | 860.8 | 929.5 | +8.0 % | 16/16 |
| 3 | 844.2 | 925.7 | +9.6 % | 16/16 |
| 4 | 840.5 | 921.2 | +9.6 % | 16/16 |
| 5 | 855.0 | 922.3 | +7.9 % | 16/16 |

Prompt ~11,000 tokens:

| step | stock pp t/s | patch pp t/s | delta | sep |
| ---: | ---: | ---: | --- | --- |
| 0 | 844.5 | 915.2 | +8.4 % | 16/16 |
| 1 | 798.7 | 901.9 | +12.9 % | 16/16 |
| 2 | 761.9 | 887.5 | +16.5 % | 16/16 |
| 3 | 739.4 | 879.6 | +19.0 % | 16/16 |
| 4 | 737.8 | 878.7 | +19.1 % | 16/16 |
| 5 | 759.3 | 885.0 | +16.6 % | 16/16 |

Stock degrades as the other slots fill (963 -> 840, -12.8 %); the patch is nearly
flat (967 -> 921, -4.8 %). The effect grows with depth, which is why I get less
than your +35 % — your 40,960 input tokens are four times deeper than my deeper
case, and my trend points that way.

With three requests running **concurrently** instead of sequentially, prefill is
+4.9 % (16/16 separation), lower because the prefills interleave and the slots
only fill during the run.

## 2. The inner-loop cost is not measurable here

Step 0 of the 5,400-token run is the control for exactly the objection in the
review: WMMA runs, same batch width as every other step, but the other slots are
empty so there is nothing to skip. That step is **+0.4 % with 7/16 separation**,
i.e. noise. The probe does not cost me anything I can measure.

Step 0 of the 11,000-token run is *not* a control (+8.4 %): at that length the
prompt already has fully masked tiles above the diagonal within its own prefill,
so the patch has something to skip even with empty slots.

Decode is unchanged: cleaned for MTP acceptance, `ms/step` is 131.31 stock vs
127.75 patched, but separation is 5/16, so that is noise. Wall-clock time per
phase is also noise (11/16). MTP acceptance was 49.8 % vs 50.6 %.

## 3. `flash_attn_mask_to_KV_max` never runs on this configuration

Before measuring I instrumented a build to scan the actual mask on the host and
count, per FA shape, how many `(jt, kb0)` tiles are entirely `-INF`. Two results
are worth reporting.

**The helper is never launched.** Its condition (`fattn-common.cuh`) is

```c
mask && K->ne[1] % FATTN_KQ_STRIDE == 0 && (Q->ne[1] >= 1024 || Q->ne[3] > 1)
```

and with `-ub 512` capping `Q->ne[1]` at 512 and `--kv-unified` keeping
`Q->ne[3] == 1`, it is false for **every** shape I observed, prefill and decode
alike. So today not even the trailing masked run is skipped here. If the intended
fix is to extend that helper, note that on this configuration it would also need
its launch condition revisited, otherwise it never gets the chance to help.

**Skippable fraction, share of all tiles:**

| streams | shape | median | max |
| ---: | --- | ---: | ---: |
| 1 | prefill | 13.9 % | 60.4 % |
| 2 | prefill | 52.2 % | 84.4 % |
| 3 | prefill | 69.2 % | 96.0 % |
| 3 | decode | 26.5 % | 87.5 % |

With one stream the *interior* skippable count is exactly 0 in all 12 shapes,
which is the control for the census itself.

Decode has far less to gain than prefill, and I think the reason is tile
granularity: a tile spans `ncols1 = 8` query rows and is only skippable if all 8
are masked. During decode those 8 rows are tokens from several slots at once, so
they are mixed and nothing is skippable; with two streams I measured interior
skip = 0 in all 9 decode shapes. During prefill the whole batch belongs to one
sequence, so entire foreign-slot blocks drop out.

## 4. Correctness, and a note on how to test it

Perplexity is bit-identical to stock, including every per-chunk value
(1.8574 +/- 0.02351, 8 chunks at `-c 4096`). `test-backend-ops -o FLASH_ATTN_EXT`
is 4574/4574 on the patched build.

For token-level identity there is a trap worth flagging. With **three concurrent**
requests, greedy, fixed prompts, the same binary run twice does **not** reproduce
itself: 1/3 streams matched, diverging at tokens 75 and 76. Batch composition
under `-np 3` depends on timing, and on this GPU batch width changes logits
(#28768). So a stock-vs-patch comparison in that shape is meaningless, and it
would have looked like the patch changing output.

Issuing the same requests **sequentially into different slots** keeps the
interior masking (idle slots are not cleared under `--no-cache-idle-slots`) while
making batch composition deterministic. In that shape the control reproduces
(5/5 steps), and stock vs patch is **identical token IDs, 5/5 steps, 640 tokens
per run**, including the steps where all three slots are full.

That is the same shape as your slots 0/1/0 validation, and it is probably why you
got identical token IDs where a concurrent test would not have.

## Summary

On gfx1201 the patch does what you claim, the direction and the depth trend match,
and the inner-loop cost the review is concerned about does not show up in my
measurements. I have not deployed it — the implementation question is yours and
the maintainer's to settle. Happy to re-run any of this against a rewrite that
uses the auxiliary kernel instead.

**Comment by JohannesGaessler** [comment]
According to the llama.cpp [AI usage policy](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md#ai-usage-policy):

>It is strictly prohibited to use AI to write your posts for you (bug reports, feature requests, pull request descriptions, Github discussions, responding to humans, ...).

**Comment by DanoPTT** [comment]
In my config (`-ub 512`, `--kv-unified`) `flash_attn_mask_to_KV_max` never runs:
the gate at `fattn-common.cuh:1108` requires `Q->ne[1] >= 1024 || Q->ne[3] > 1`,
but `-ub 512` caps `Q->ne[1]` at 512 and `--kv-unified` keeps `Q->ne[3] == 1`.
So `KV_max` stays null and the kernel never skips a single tile.
Verified against current master (`60081bb2`).

**Comment by moristeven477-ship-it** [comment]
https://github.com/moristeven477-ship-it/llama.cpp/releases/tag/rdna3-masked-kv-v2-20260921

**Comment by DanoPTT** [comment]
As a happy user of the V1 patch, I tested the V2 patch on gfx1201 (Radeon AI PRO R9700, ROCm 10.0.0, Windows), so this is
a different WMMA generation than the gfx1100 results above.
llama-server, Qwen3.8-27B, Q5_K_XL, `-np 3 --kv-unified -ub 512 -c 161280 --no-cache-idle-slots`, 
sequential requests filling slots 0, 1, 2, 0, 1, 2 so the mask gets interior -INF blocks while each request still runs alone. 
Three builds off one tree, differing only in the patch: no skip, V1, V2. 
8 visits in A,B,B,A,A,B,B,A order, so each median below is over the 4 visits of that arm.

Prompt processing, t/s. Step 0 has the other slots empty, so there is nothing to skip; by step 2 all three slots hold content.

| step | ~5.4k tokens |  |  | ~11k tokens |  |  |
|---|---|---|---|---|---|---|
|  | no skip | V1 | V2 | no skip | V1 | V2 |
| 0 | 894.3 | 892.0 | 901.0 | 783.6 | 841.6 | 870.7 |
| 1 | 846.5 | 874.2 | 887.5 | 746.6 | 826.9 | 864.6 |
| 2 | 803.2 | 857.0 | 884.0 | 712.2 | 813.1 | 858.5 |
| 3 | 785.3 | 849.1 | 880.9 | 690.4 | 804.9 | 855.1 |
| 4 | 780.6 | 847.8 | 877.7 | 688.5 | 804.3 | 853.7 |
| 5 | 795.4 | 848.7 | 876.2 | 708.6 | 810.1 | 855.5 |

Two runs against the same no-skip build (V2 in one, V1 in the other); its control column agreed between them within 0.5%.
Over steps 2-5 that is +10.1-12.4% for V2 against no skip at 5.4k and +20.5-24.0% at 11k, 
of which +2.4-3.1% and +5.0-5.8% is V2 over V1.
More useful than the percentages is the shape: across the steps prompt processing falls 12.1% without the patch, 4.4% with V1 and 2.0% with V2 (11k column), so V2 largely removes the decay reported in #28495. 

Worth noting for this config: the separate `use_mask_blocks` path is not behind the `Q->ne[1] >= 1024 || Q->ne[3] > 1` condition, so it actually runs here. 
As I mentioned earlier in this thread, that condition means `flash_attn_mask_to_KV_max` never fires with `-ub 512` and `--kv-unified`, so on this setup the existing helper skips nothing at all, not even the mask tail.

Probe cost: step 0 is the case with the other slots empty, where there is least to gain, and V2 is not slower there in either column.

Output is unchanged: perplexity is bit-identical and token IDs match over 640 tokens in the deterministic sequential-slot shape. test-backend-ops FLASH_ATTN_EXT passes 4599/4599 including the new mask-class cases.

Important note: this tree has no sparse FA (#27970), so I dropped the use_sparse template parameter and the indices arguments when applying the patch. The numbers are for that adaptation, not for the patch verbatim.

**Comment by JohannesGaessler** [comment]
The way this is implemented still introduces too much technical debt. This needs a human to do the software design.

**Comment by moristeven477-ship-it** [comment]
> The way this is implemented still introduces too much technical debt. This needs a human to do the software design.这种实现方式仍然引入了过多的技术负债。这需要有人来完成软件设计。

Can you give me some hint or target please?

**Comment by JohannesGaessler** [comment]
To make it clear: my impression is that the design in this PR is mostly driven by a language model and significantly more human effort is needed to come up with a better design. This should be done by someone else.
